### PR TITLE
Extract mixins from Snippet views and use it in generic create/edit/delete views

### PR DESCRIPTION
**Pull request by laymonage**
~~Make sure #8332, #8335, #8344, #8346, #8351, and #8352 are merged first. For now, I've merged them all together so I can test it locally.~~

_Please check the following:_

- [x] Do the tests still pass?[^1]
- [x] Does the code comply with the style guide? 
    - [x] Run `make lint` from the Wagtail root. 
- [ ] For Python changes: Have you added tests to cover the new/fixed behaviour?
- [ ] For front-end changes: Did you test on all of Wagtail’s supported environments?[^2]
    - [ ] **Please list the exact browser and operating system versions you tested**:
    - [ ] **Please list which assistive technologies [^3] you tested**: 
- [ ] For new features: Has the documentation been updated accordingly?

**Please describe additional details for testing this change**. 

[^1]: [Development Testing](https://docs.wagtail.org/en/latest/contributing/developing.html#testing)
[^2]: [Browser and device support](https://docs.wagtail.org/en/latest/contributing/developing.html#browser-and-device-support)
[^3]: [Accessibility Target](https://docs.wagtail.org/en/latest/contributing/developing.html#accessibility-targets)

### CODE REVIEW

File: wagtail/admin/views/generic/models.py [hunk 1]
```diff
@@ -63,6 +69,51 @@ class IndexView(PermissionCheckedMixin, WagtailAdminTemplateMixin, BaseListView)
...
     page_kwarg = "p"
     default_ordering = None
+    is_searchable = None
+    search_kwarg = "q"
+
+    def setup(self, request, *args, **kwargs):
+        super().setup(request, *args, **kwargs)
+        self.setup_search()
+
+    def setup_search(self):
```
Comment (laymonage) [inline wagtail/admin/views/generic/models.py:79]: Not turning this into a mixin yet as I don't think it's necessary

File: wagtail/admin/views/generic/mixins.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+from wagtail.models import Locale, TranslatableMixin
+
+
+class LocaleMixin:
+    def setup(self, request, *args, **kwargs):
+        super().setup(request, *args, **kwargs)
+        self.locale = self.get_locale()
+
+    def get_translations_context(self):
```
Comment (laymonage) [inline wagtail/admin/views/generic/mixins.py:12]: Or maybe just `get_translations(self)`?

File: wagtail/admin/views/generic/hooks.py [hunk 3]
```diff
@@ -16,3 +15,52 @@ def run_hook(self, hook_name, *args, **kwargs):
             result = fn(*args, **kwargs)
             if hasattr(result, "status_code"):
                 return result
+        return None
+
+
+class BeforeAfterHookMixin(HookResponseMixin):
```
Comment (laymonage) [inline wagtail/admin/views/generic/hooks.py:21]: Should we move these to `mixins.py`?

File: wagtail/snippets/views/snippets.py [hunk 4]
```diff
@@ -360,94 +289,65 @@ def get_form_kwargs(self):
...
+        super().setup(request, *args, **kwargs)
 
     def _get_model(self):
         return get_snippet_model_from_url_params(self.app_label, self.model_name)
 
     def _get_edit_handler(self):
         return get_snippet_edit_handler(self.model)
 
-    def _get_locale(self):
+    def get_locale(self):
```
Comment (laymonage) [inline wagtail/snippets/views/snippets.py:350]: Should this logic (i.e. retrieving the locale from `self.object.locale`) be included in `LocaleMixin`? If so, which one takes precedence (compared to `?locale=` query param)?

File: wagtail/admin/views/generic/hooks.py (hunk 3 above)
Comment (kaedroho) [reply wagtail/admin/views/generic/hooks.py:21]: yep, I think so

File: wagtail/admin/views/generic/mixins.py (hunk 2 above)
Comment (kaedroho) [reply wagtail/admin/views/generic/mixins.py:13]: Should we rename this to ``get _locales`` and make it return a list/queryset of locales to display in the dropdown?

It feels like the formatting of the locale data for the template context belongs in ``get_context_data``?

**Review by kaedroho — APPROVED** [review]
This all looks good to me, thanks!

**Pull request merged**